File: DFS_BFS/549-findEvenNumber-bfs.py
from collections import deque,Counter
from copy import deepcopy
def findEvenNumbers(digits):
    C=Counter(digits)
    ans=set()
    q=deque([(0,C)])
    while q:
        x,c=q.popleft()
        if x>99:
            if (x&1)==0:
                ans.add(x)
        else:
            for i in range(10):
                cc=deepcopy(c)
                if cc[i]>0:
                    cc[i]-=1
                    q.append((x*10+i,cc))
    return sorted(ans)

nums=[2,1,3,0,0]
print(findEvenNumbers(nums))

File: DFS_BFS/549-findEvenNumber-bfs.cpp
#include<iostream>
#include<set>
#include<vector>
#include<deque>
#include<map>

using std::cout;
using VEC=std::vector<int>;
using MAP=std::map<int,int>;
using SET=std::set<int>;


MAP myCounter(const VEC& digits)
{
    MAP ans{};
    for(auto &i:digits)
    {
        ans[i]+=1;
    }
    return ans;
}

SET findEvenNumbers(const VEC& digits)
{
    MAP C=myCounter(digits);
    SET ans{};
    std::deque<std::pair<int,MAP>> q{};
    std::pair<int,MAP> p(0,C);
    q.push_back(p);
    while(!q.empty())
    {
        std::pair cur=q.front();
        q.pop_front();
        int x=cur.first;
        MAP c=cur.second;
        if(x>99)
        {
            if((x&1)==0)
                ans.insert(x);
        }
        else
            for(int i=0;i<10;++i)
            {
                if(c[i]>0)
                {
                    MAP cc{c};
                    cc[i]-=1;
                    std::pair<int,MAP> tmp{x*10+i,cc};
                    q.push_back(tmp);
                }
            }
    }
    return ans;
}


int main()
{
    VEC nums{1,2,3,0};
    SET res{findEvenNumbers(nums)};
    for(auto i:res)
        cout<<i<<',';
}